### PR TITLE
StringHelper: Use functional programming to improve code readability #7838

### LINKED ISSUE #7838: StringHelper: Use functional programming to improve code readability
Many methods in StringHelper use loops. They can be simplified to use Java 8 Streams instead.

### PR DESCRIPTION
**Pull request by darrenwee**
Fixes #7838

**Outline of Solution**

I refactored methods to use Java 8's streams API to improve readability. I left some functions untouched because I felt refactoring them would not improve the code quality or I wasn't sure how to refactor them to use streams. Examples of those methods are [`csvToHtmlTable`](https://github.com/darrenwee/teammates/blob/84b3b649d5ac723acf1b2d079ff9e35cd15f75bd/src/main/java/teammates/common/util/StringHelper.java#L332) and [`byteArrayToHexString`](https://github.com/darrenwee/teammates/blob/84b3b649d5ac723acf1b2d079ff9e35cd15f75bd/src/main/java/teammates/common/util/StringHelper.java#L307).

I see #8257 was also filed two hours ago. I posted in #7838 that I would take the issue. Possible to merge the two together?

(sorry for the delay, I'm on vacation and don't have consistent access to my laptop)

### CODE REVIEW

**Comment by thenaesh** [comment]
@darrenwee Let me know when this is ready for review. As for #8257, since both PRs are touching the same functions in the same file it might be difficult to merge both change sets together.

**Comment by darrenwee** [comment]
@thenaesh Ready for review. Thanks.

I think you can merge this one after review, then re-evaluate the other PR against the new `master`? Not sure what the correct protocol is.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 1]
```diff
@@ -429,14 +411,12 @@ private static String handleNewLine(String str) {
...
-                return false;
-            }
+        if (rowData.isEmpty()) {
+            return true;
         }
 
-        return true;
+        return rowData.stream()
+                .anyMatch(r -> r.isEmpty());
     }
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: Have another look at this one. Doesn't look like the new code is same as previous.

Comment (thenaesh) [reply src/main/java/teammates/common/util/StringHelper.java]: @darrenwee This should use `allMatch` and not `anyMatch`. If even a single element is nonempty this method returns false, so this method returns true exactly when all elements are empty.

@damithc Thanks for the catch, I missed this one out.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 2]
```diff
@@ -312,34 +297,31 @@ public static String replaceIllegalChars(String str, String regex, char replacem
...
+        IntStream.range(0, charArray.length)
+                .filter(i -> !isMatching(Character.toString(charArray[i]), regex))
+                .forEach(i -> charArray[i] = replacement);
 
         return String.valueOf(charArray);
     }
 
     public static String byteArrayToHexString(byte[] bytes) {
         StringBuilder sb = new StringBuilder(bytes.length * 2);
+
```
Comment (thenaesh) [inline src/main/java/teammates/common/util/StringHelper.java]: Extraneous change, please remove.

File: src/main/java/teammates/common/util/StringHelper.java [hunk 3]
```diff
@@ -312,34 +297,31 @@ public static String replaceIllegalChars(String str, String regex, char replacem
...
         for (byte b : bytes) {
             int v = b & 0xff;
             if (v < 16) {
                 sb.append('0');
             }
             sb.append(Integer.toHexString(v));
         }
+
```
Comment (thenaesh) [inline src/main/java/teammates/common/util/StringHelper.java]: This too.

File: src/main/java/teammates/common/util/StringHelper.java (hunk 1 above)
Comment (thenaesh) [reply src/main/java/teammates/common/util/StringHelper.java]: Also, the fact that Travis passes means that we don't have a test for this...

**Comment by damithc** [comment]
For our peace of mind,  I suggest we add any missing tests in this PR too so that all methods touched in this PR are covered by tests. If the method is private, we can cover the public methods using that private method.

File: src/main/java/teammates/common/util/StringHelper.java (hunk 1 above)
Comment (darrenwee) [reply src/main/java/teammates/common/util/StringHelper.java]: Oops, accidentally inverted this in my head. Yes, it should be the complement of `anyMatch`.

**Comment by darrenwee** [comment]
Hey @thenaesh, this is ready for review again. I have added some test cases for some of the functions I modified, but the current test cases were pretty comprehensive. The missing test case for that `anyMatch` issue was resolved by another PR. I also fixed issues raised in your review.

There is one test case I added which I am unsure of. This (from `testJoinWithListOfIntegers`) passes:
```
assertEquals("5||14||null", StringHelper.join("||", Arrays.asList(5, 14, null)));
```
If you feed a `null` object to the list of integers, the resulting string from joining the list contains `null` as a string. I am not sure if this is desired behavior.

**Comment by darrenwee** [comment]
@thenaesh good to go!

File: src/main/java/teammates/common/util/StringHelper.java [hunk 4]
```diff
@@ -87,11 +83,8 @@ public static String truncate(String inputString, int truncateLength) {
         if (inputString.length() <= truncateLength) {
             return inputString;
         }
-        String result = inputString;
-        if (inputString.length() > truncateLength) {
-            result = inputString.substring(0, truncateLength - 3) + "...";
-        }
-        return result;
+
+        return inputString.substring(0, truncateLength - 3) + "...";
```
Comment (damithc) [inline src/main/java/teammates/common/util/StringHelper.java]: Does this work even if the inputString is shorter than `truncateLength`?

Comment (darrenwee) [reply src/main/java/teammates/common/util/StringHelper.java]: Yes, but only if `inputString` is at least `truncateLength-3` characters long. That case is caught by the if-block before the return statement.

**Comment by darrenwee** [comment]
Bump!

**Comment by damithc** [comment]
@thenaesh for merging.
@darrenwee FYI, I've updated the branch (via GitHub UI) as it has gone out of date. i.e., your local is behind your fork now.

**Pull request merged**